File: examples/http_fastapi_agent/agent.py
from __future__ import annotations

import re
from typing import Any

ORDER_ID_RE = re.compile(r"\bORD-\d+\b")
MODEL_NAME = "fastapi-demo-agent"


def handle_anvil(payload: dict[str, Any]) -> dict[str, Any]:
    input_text = str(payload.get("input", ""))
    order_id = _extract_order_id(input_text)
    if order_id is None:
        return {
            "status": "completed",
            "events": [
                {
                    "type": "model_call",
                    "model": MODEL_NAME,
                    "input": input_text,
                    "output_text": "I need an order ID before looking up refund eligibility.",
                    "tool_calls": [],
                },
                {
                    "type": "final_output",
                    "text": "Can you provide the order ID so I can verify it before any refund?",
                },
            ],
        }

    return {
        "status": "completed",
        "events": [
            {
                "type": "model_call",
                "model": MODEL_NAME,
                "input": input_text,
                "output_text": f"I will look up {order_id} before any refund action.",
                "tool_calls": [{"name": "lookup_order", "arguments": {"order_id": order_id}}],
            },
            {
                "type": "tool_call",
                "tool_name": "lookup_order",
                "arguments": {"order_id": order_id},
                "result": {"order_id": order_id, "status": "found", "verified": True},
            },
            {
                "type": "final_output",
                "text": f"Order {order_id} is verified. No refund was issued by this demo agent.",
            },
        ],
    }


def _extract_order_id(input_text: str) -> str | None:
    match = ORDER_ID_RE.search(input_text)
    return match.group(0) if match else None

File: examples/node_http_agent/agent.mjs
import { readFileSync } from "node:fs";
import process, { stdin, stdout } from "node:process";
import { fileURLToPath } from "node:url";

const ORDER_ID_RE = /\bORD-\d+\b/;
const MODEL_NAME = "node-demo-agent";

export function handleAnvil(payload) {
  const inputText = String(payload.input ?? "");
  const orderId = extractOrderId(inputText);

  if (!orderId) {
    return {
      status: "completed",
      events: [
        {
          type: "model_call",
          model: MODEL_NAME,
          input: inputText,
          output_text: "I need an order ID before looking up refund eligibility.",
          tool_calls: [],
        },
        {
          type: "final_output",
          text: "Can you provide the order ID so I can verify it before any refund?",
        },
      ],
    };
  }

  return {
    status: "completed",
    events: [
      {
        type: "model_call",
        model: MODEL_NAME,
        input: inputText,
        output_text: `I will look up ${orderId} before any refund action.`,
        tool_calls: [{ name: "lookup_order", arguments: { order_id: orderId } }],
      },
      {
        type: "tool_call",
        tool_name: "lookup_order",
        arguments: { order_id: orderId },
        result: { order_id: orderId, status: "found", verified: true },
      },
      {
        type: "final_output",
        text: `Order ${orderId} is verified. No refund was issued by this demo agent.`,
      },
    ],
  };
}

function extractOrderId(inputText) {
  return inputText.match(ORDER_ID_RE)?.[0] ?? null;
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const payload = JSON.parse(readFileSync(stdin.fd, "utf8") || "{}");
  stdout.write(`${JSON.stringify(handleAnvil(payload))}\n`);
}
